fdbserver/DataDistribution.actor.cpp

```diff
@@ -3438,13 +3455,20 @@ ACTOR Future<Void> storageServerTracker(
 
 			// If the storage server is in the excluded servers list, it is undesired
 			NetworkAddress a = server->lastKnownInterface.address();
-			AddressExclusion addr( a.ip, a.port );
-			AddressExclusion ipaddr( a.ip );
+			state AddressExclusion addr( a.ip, a.port );
+			state AddressExclusion ipaddr( a.ip );
 			if (self->excludedServers.get( addr ) || self->excludedServers.get( ipaddr )) {
 				TraceEvent(SevWarn, "UndesiredStorageServer", self->distributorId).detail("Server", server->id)
 					.detail("Excluded", self->excludedServers.get( addr ) ? addr.toString() : ipaddr.toString());
 				status.isUndesired = true;
 				status.isWrongConfiguration = true;
+				if (self->failedServers.find(addr) != self->failedServers.end() || self->failedServers.find(ipaddr) != self->failedServers.end()) {
+					TraceEvent(SevWarn, "FailedServerRemoveKeys", self->distributorId)
+					    .detail("Address", addr.toString())
+					    .detail("ServerID", server->id);
+					wait(removeKeysFromFailedServer(cx, server->id, self->lock));
+					self->shardsAffectedByTeamFailure->eraseServer(server->id);
```

Comment:
Can we buggify this by adding a random wait()? I want to confirm if it is valid for shardsAffectedByTeamFailure to have data for the server while the server's ownership for shards have been revoked in removeKeysFromFailedServer?

Consider adding `if (BUGGIFY) wait(delay(some_random_time));` just before `self->shardsAffectedByTeamFailure->eraseServer(server->id);`.


Another question:
Can we move `self->shardsAffectedByTeamFailure->eraseServer(server->id);` before the `wait`? 
If it can, it may reduce the useless work in queuing data to the failed server while we are in the process of removing the failed server.

Reply:
I was a little confused on the exact ordering you wanted here. Can you check the latest commit to see if that's what you wanted?

Reply:
I ran some tests and I believe that we need to keep `self->shardsAffectedByTeamFailure->eraseServer(server->id);` after `wait(removeKeysFromFailedServer(cx, server->id, self->lock));` If the order were reversed, shards could be assigned to the tracker in a race since they haven't been erased from system keyspace yet, which would delay the removal of the storage server.